doc/test: several `HelpExampleRpc` curl examples emit invalid JSON; add a regression check to rpc_help.py

`HelpExampleRpc` interpolates its `args` string directly into a JSON-RPC payload
(src/rpc/util.cpp:201-204) with no validation that the result is a well-formed
JSON array body. On master (556988790a), 13 of the 138 literal `HelpExampleRpc`
call sites produce a payload that fails JSON parsing, so the documented curl
command is unusable as printed.

Three failure modes:

**Missing comma separators** (CLI-style arg string reused in the RPC example):
- src/rpc/blockchain.cpp:543 `getblockfrompeer`
- src/rpc/net.cpp:338 `addnode`
- src/rpc/net.cpp:413 `addconnection`
- src/rpc/net.cpp:1086 `sendmsgtopeer`
- src/wallet/rpc/backup.cpp:628 `restorewallet`

`addnode` shows this is an oversight rather than style — it has the first comma
and is missing only the second:

    HelpExampleCli("addnode", "\"192.168.0.6:8333\" \"onetry\" true")
  + HelpExampleRpc("addnode", "\"192.168.0.6:8333\", \"onetry\" true")

renders `"params": ["192.168.0.6:8333", "onetry" true]`
→ `Expecting ',' delimiter: line 1 column 99`

**Bare unquoted strings** (valid as a CLI placeholder, a syntax error in JSON):
src/rpc/mempool.cpp:871 `getmempoolcluster`, src/rpc/mempool.cpp:1171
`importmempool`, src/rpc/node.cpp:381 `getindexinfo`,
src/wallet/rpc/addresses.cpp:646 `listlabels`, src/wallet/rpc/wallet.cpp:454
`unloadwallet`, src/wallet/rpc/wallet.cpp:774 `createwalletdescriptor`,
src/wallet/rpc/wallet.cpp:864 `addhdkey`.

**Invalid backslash escape** — src/wallet/rpc/wallet.cpp:246 `loadwallet`,
the "absolute path (Windows)" example renders
`"params": ["DriveLetter:\path\to\walletname\"]` → `Invalid \escape`. The
trailing `\"` also escapes the closing quote. The sibling CLI example on line
245 is correct as-is; only the JSON form needs the backslashes doubled.

### Why file this now rather than another round of manual fixes

In #31275 sipa ran the runtime version of this check ("I wanted to verify if all
invocations of `HelpExampleRpc` produce valid JSON, so I tested by adding a
`UniValue::read` inside the function", 2025-11-10) and listed 11 failing RPCs.
That list is an exact subset of the 13 above. Since then:

| RPC | landed on master | vs. the 2025-11-10 audit |
|---|---|---|
| `getmempoolcluster` | 2025-11-25 (`fa283d28e2`, #33629) | +15 days |
| `addhdkey` | 2026-05-14 (`cad5f56045`, #29136) | +6 months |

None of the original 11 were fixed, two more were added, and #31275 was closed
unmerged on 2026-03-14 — `addhdkey`'s example landed two months after that. Six
separate PRs have fixed instances of this one-by-one over the years and the
population still grows, because nothing enforces it.

### Proposed direction

1. **Add the guard first.** `test/functional/rpc_help.py::dump_help()` already
   iterates every registered command and calls `self.nodes[0].help(call)`.
   Extend it to extract each `> curl ... --data-binary '<payload>'` line from
   the help text and assert `json.loads(payload)` succeeds. Roughly ten lines,
   no new test file, covers every current and future RPC, and would have caught
   all 13 — including the ~16 call sites whose args are built from expressions
   (`EXAMPLE_ADDRESS`, `strprintf`, …) that a static scan has to skip. This is
   sipa's second suggestion from #31275.
2. **Then fix the 13**: add the missing commas, quote the bare placeholders,
   and double the backslashes in the `loadwallet` Windows example.

Worth settling direction before anyone writes a patch: sipa's first suggestion
was to drop the curl examples entirely, which would moot the fixes. Happy to
open a PR for whichever the maintainers prefer.

Tried implementing the dump_help() extension suggested above (regex-extract
the curl --data-binary payload, json.loads() it) and ran it against current
master. It caught getblockfrompeer on the first run  same missing-comma
pattern as the addnode example, different call site:

AssertionError: HelpExampleRpc for 'getblockfrompeer' is not valid JSON:
Expecting ',' delimiter: line 1 column 146 (char 145)

So the approach checks out independent of the 13 already listed above.

--- a/src/rpc/blockchain.cpp
+++ b/src/rpc/blockchain.cpp
@@ -540,7 +540,7 @@ static RPCMethod getblockfrompeer()
         RPCResult{RPCResult::Type::OBJ, "", /*optional=*/false, "", {}},
         RPCExamples{
             HelpExampleCli("getblockfrompeer", "\"00000000c937983704a73af28acdec37b049d214adbda81d7e2a3dd146f6ed09\" 0")
-            + HelpExampleRpc("getblockfrompeer", "\"00000000c937983704a73af28acdec37b049d214adbda81d7e2a3dd146f6ed09\" 0")
+            + HelpExampleRpc("getblockfrompeer", R"("00000000c937983704a73af28acdec37b049d214adbda81d7e2a3dd146f6ed09", 0)")
         },
         [](const RPCMethod& self, const JSONRPCRequest& request) -> UniValue
 {

--- a/src/rpc/mempool.cpp
+++ b/src/rpc/mempool.cpp
@@ -868,7 +868,7 @@ static RPCMethod getmempoolcluster()
             RPCResult::Type::OBJ, "", "", ClusterDescription()},
         RPCExamples{
             HelpExampleCli("getmempoolcluster", "txid")
-            + HelpExampleRpc("getmempoolcluster", "txid")
+            + HelpExampleRpc("getmempoolcluster", R"("txid")")
         },
         [](const RPCMethod& self, const JSONRPCRequest& request) -> UniValue
 {
@@ -1168,7 +1168,7 @@ static RPCMethod importmempool()
              RPCArgOptions{.oneline_description = "options"}},
         },
         RPCResult{RPCResult::Type::OBJ, "", "", std::vector<RPCResult>{}},
-        RPCExamples{HelpExampleCli("importmempool", "/path/to/mempool.dat") + HelpExampleRpc("importmempool", "/path/to/mempool.dat")},
+        RPCExamples{HelpExampleCli("importmempool", "/path/to/mempool.dat") + HelpExampleRpc("importmempool", R"("/path/to/mempool.dat")")},
         [](const RPCMethod& self, const JSONRPCRequest& request) -> UniValue {
             const NodeContext& node{EnsureAnyNodeContext(request.context)};
 

--- a/src/rpc/net.cpp
+++ b/src/rpc/net.cpp
@@ -335,7 +335,7 @@ static RPCMethod addnode()
                 RPCResult{RPCResult::Type::NONE, "", ""},
                 RPCExamples{
                     HelpExampleCli("addnode", "\"192.168.0.6:8333\" \"onetry\" true")
-            + HelpExampleRpc("addnode", "\"192.168.0.6:8333\", \"onetry\" true")
+            + HelpExampleRpc("addnode", R"("192.168.0.6:8333", "onetry", true)")
                 },
         [](const RPCMethod& self, const JSONRPCRequest& request) -> UniValue
 {
@@ -410,7 +410,7 @@ static RPCMethod addconnection()
             }},
         RPCExamples{
             HelpExampleCli("addconnection", "\"192.168.0.6:8333\" \"outbound-full-relay\" true")
-            + HelpExampleRpc("addconnection", "\"192.168.0.6:8333\" \"outbound-full-relay\" true")
+            + HelpExampleRpc("addconnection", R"("192.168.0.6:8333", "outbound-full-relay", true)")
         },
         [](const RPCMethod& self, const JSONRPCRequest& request) -> UniValue
 {
@@ -1083,7 +1083,7 @@ static RPCMethod sendmsgtopeer()
         },
         RPCResult{RPCResult::Type::OBJ, "", "", std::vector<RPCResult>{}},
         RPCExamples{
-            HelpExampleCli("sendmsgtopeer", "0 \"addr\" \"ffffff\"") + HelpExampleRpc("sendmsgtopeer", "0 \"addr\" \"ffffff\"")},
+            HelpExampleCli("sendmsgtopeer", "0 \"addr\" \"ffffff\"") + HelpExampleRpc("sendmsgtopeer", R"(0, "addr", "ffffff")")},
         [](const RPCMethod& self, const JSONRPCRequest& request) -> UniValue {
             const NodeId peer_id{request.params[0].getInt<int64_t>()};
             const auto msg_type{self.Arg<std::string_view>("msg_type")};

--- a/src/rpc/node.cpp
+++ b/src/rpc/node.cpp
@@ -378,7 +378,7 @@ static RPCMethod getindexinfo()
                     HelpExampleCli("getindexinfo", "")
                   + HelpExampleRpc("getindexinfo", "")
                   + HelpExampleCli("getindexinfo", "txindex")
-                  + HelpExampleRpc("getindexinfo", "txindex")
+                  + HelpExampleRpc("getindexinfo", R"("txindex")")
                 },
                 [](const RPCMethod& self, const JSONRPCRequest& request) -> UniValue
 {

--- a/src/wallet/rpc/addresses.cpp
+++ b/src/wallet/rpc/addresses.cpp
@@ -643,7 +643,7 @@ RPCMethod listlabels()
             "\nList labels that have sending addresses\n"
             + HelpExampleCli("listlabels", "send") +
             "\nAs a JSON-RPC call\n"
-            + HelpExampleRpc("listlabels", "receive")
+            + HelpExampleRpc("listlabels", R"("receive")")
                 },
         [](const RPCMethod& self, const JSONRPCRequest& request) -> UniValue
 {

--- a/src/wallet/rpc/backup.cpp
+++ b/src/wallet/rpc/backup.cpp
@@ -625,9 +625,9 @@ RPCMethod restorewallet()
         },
         RPCExamples{
             HelpExampleCli("restorewallet", "\"testwallet\" \"home\\backups\\backup-file.bak\"")
-            + HelpExampleRpc("restorewallet", "\"testwallet\" \"home\\backups\\backup-file.bak\"")
-            + HelpExampleCliNamed("restorewallet", {{"wallet_name", "testwallet"}, {"backup_file", "home\\backups\\backup-file.bak\""}, {"load_on_startup", true}})
-            + HelpExampleRpcNamed("restorewallet", {{"wallet_name", "testwallet"}, {"backup_file", "home\\backups\\backup-file.bak\""}, {"load_on_startup", true}})
+            + HelpExampleRpc("restorewallet", R"("testwallet", "home\\backups\\backup-file.bak")")
+            + HelpExampleCliNamed("restorewallet", {{"wallet_name", "testwallet"}, {"backup_file", "home\\backups\\backup-file.bak"}, {"load_on_startup", true}})
+            + HelpExampleRpcNamed("restorewallet", {{"wallet_name", "testwallet"}, {"backup_file", "home\\backups\\backup-file.bak"}, {"load_on_startup", true}})
         },
         [](const RPCMethod& self, const JSONRPCRequest& request) -> UniValue
 {

--- a/src/wallet/rpc/coins.cpp
+++ b/src/wallet/rpc/coins.cpp
@@ -514,7 +514,7 @@ RPCMethod listunspent()
                 RPCExamples{
                     HelpExampleCli("listunspent", "")
             + HelpExampleCli("listunspent", "6 9999999 \"[\\\"" + EXAMPLE_ADDRESS[0] + "\\\",\\\"" + EXAMPLE_ADDRESS[1] + "\\\"]\"")
-            + HelpExampleRpc("listunspent", "6, 9999999 \"[\\\"" + EXAMPLE_ADDRESS[0] + "\\\",\\\"" + EXAMPLE_ADDRESS[1] + "\\\"]\"")
+            + HelpExampleRpc("listunspent", strprintf(R"(6, 9999999, ["%s","%s"])", EXAMPLE_ADDRESS[0], EXAMPLE_ADDRESS[1]))
             + HelpExampleCli("listunspent", "6 9999999 '[]' true '{ \"minimumAmount\": 0.005 }'")
             + HelpExampleRpc("listunspent", "6, 9999999, [] , true, { \"minimumAmount\": 0.005 } ")
                 },

--- a/src/wallet/rpc/coins.cpp
+++ b/src/wallet/rpc/coins.cpp
@@ -514,7 +514,7 @@ RPCMethod listunspent()
                 RPCExamples{
                     HelpExampleCli("listunspent", "")
             + HelpExampleCli("listunspent", "6 9999999 \"[\\\"" + EXAMPLE_ADDRESS[0] + "\\\",\\\"" + EXAMPLE_ADDRESS[1] + "\\\"]\"")
-            + HelpExampleRpc("listunspent", "6, 9999999 \"[\\\"" + EXAMPLE_ADDRESS[0] + "\\\",\\\"" + EXAMPLE_ADDRESS[1] + "\\\"]\"")
+            + HelpExampleRpc("listunspent", "6, 9999999, \"[\\\"" + EXAMPLE_ADDRESS[0] + "\\\",\\\"" + EXAMPLE_ADDRESS[1] + "\\\"]\"")
             + HelpExampleCli("listunspent", "6 9999999 '[]' true '{ \"minimumAmount\": 0.005 }'")
             + HelpExampleRpc("listunspent", "6, 9999999, [] , true, { \"minimumAmount\": 0.005 } ")
                 },

--- a/src/wallet/rpc/wallet.cpp
+++ b/src/wallet/rpc/wallet.cpp
@@ -243,7 +243,7 @@ static RPCMethod loadwallet()
                     + HelpExampleRpc("loadwallet", "\"/path/to/walletname/\"")
                     + "\nLoad wallet using absolute path (Windows):\n"
                     + HelpExampleCli("loadwallet", "\"DriveLetter:\\path\\to\\walletname\\\"")
-                    + HelpExampleRpc("loadwallet", "\"DriveLetter:\\path\\to\\walletname\\\"")
+                    + HelpExampleRpc("loadwallet", R"("DriveLetter:\\path\\to\\walletname")")
                 },
         [](const RPCMethod& self, const JSONRPCRequest& request) -> UniValue
 {
@@ -451,7 +451,7 @@ static RPCMethod unloadwallet()
                 }},
                 RPCExamples{
                     HelpExampleCli("unloadwallet", "wallet_name")
-            + HelpExampleRpc("unloadwallet", "wallet_name")
+            + HelpExampleRpc("unloadwallet", R"("wallet_name")")
                 },
         [](const RPCMethod& self, const JSONRPCRequest& request) -> UniValue
 {
@@ -771,7 +771,7 @@ static RPCMethod createwalletdescriptor()
         },
         RPCExamples{
             HelpExampleCli("createwalletdescriptor", "bech32m")
-            + HelpExampleRpc("createwalletdescriptor", "bech32m")
+            + HelpExampleRpc("createwalletdescriptor", R"("bech32m")")
         },
         [](const RPCMethod& self, const JSONRPCRequest& request) -> UniValue
         {
@@ -861,7 +861,7 @@ RPCMethod addhdkey()
             },
         },
         RPCExamples{
-            HelpExampleCli("addhdkey", "xprv") + HelpExampleRpc("addhdkey", "xprv")
+            HelpExampleCli("addhdkey", "xprv") + HelpExampleRpc("addhdkey", R"("xprv")")
         },
         [&](const RPCMethod& self, const JSONRPCRequest& request) -> UniValue
         {

--- a/test/functional/rpc_help.py
+++ b/test/functional/rpc_help.py
@@ -8,6 +8,7 @@ from test_framework.test_framework import BitcoinTestFramework
 from test_framework.util import assert_equal, assert_raises_rpc_error
 
 from collections import defaultdict
+import json
 import os
 import re
 
@@ -153,9 +154,17 @@ class HelpRpcTest(BitcoinTestFramework):
         os.mkdir(dump_dir)
         calls = [line.split(' ', 1)[0] for line in self.nodes[0].help().splitlines() if line and not line.startswith('==')]
         for call in calls:
+            help_text = self.nodes[0].help(call)
             with open(os.path.join(dump_dir, call), 'w') as f:
                 # Make sure the node can generate the help at runtime without crashing
-                f.write(self.nodes[0].help(call))
+                f.write(help_text)
+            # Make sure any curl examples have a JSON-RPC payload that is valid JSON
+            for match in re.finditer(r"--data-binary '(.*)' -H", help_text):
+                payload = match.group(1)
+                try:
+                    json.loads(payload)
+                except json.JSONDecodeError as e:
+                    raise AssertionError(f"HelpExampleRpc for '{call}' is not valid JSON: {payload!r}\n{e}")
 
     def wallet_help(self):
         assert 'getnewaddress ( "label" "address_type" )' in self.nodes[0].help('getnewaddress')

